File: src/android/ad_mob/src/main/java/com/ee/internal/AdMobAppOpenAd.kt
package com.ee.internal

import android.app.Activity
import android.app.Application
import androidx.annotation.AnyThread
import com.ee.*
import com.google.android.gms.ads.*
import com.google.android.gms.ads.appopen.AppOpenAd
import com.google.android.gms.ads.appopen.AppOpenAd.AppOpenAdLoadCallback
import kotlinx.serialization.Serializable
import java.util.concurrent.atomic.AtomicBoolean

internal class AdMobAppOpenAd(
    private val _bridge: IMessageBridge,
    private val _logger: ILogger,
    private val _application: Application,
    private var _activity: Activity?,
    private val _adId: String,
    private val _onAdPaid: (AdPaidResponse) -> Unit
) : IFullScreenAd {
    @Serializable
    @Suppress("unused")
    private class ErrorResponse(
        val code: Int,
        val message: String
    )

    companion object {
        private val kTag = AdMobAppOpenAd::class.java.name
    }

    private val _messageHelper = MessageHelper("AdMobAppOpenAd", _adId)
    private val _helper = FullScreenAdHelper(_bridge, this, _messageHelper)
    private val _isLoaded = AtomicBoolean(false)
    private var _ad: AppOpenAd? = null

    init {
        _logger.info("$kTag: constructor: adId = $_adId")
        registerHandlers()
    }

    override fun onCreate(activity: Activity) {
        _activity = activity
    }

    override fun onResume() {
    }

    override fun onPause() {
    }

    override fun onDestroy() {
        _activity = null
    }

    override fun destroy() {
        _logger.info("$kTag: destroy: adId = $_adId")
        deregisterHandlers()
    }

    @AnyThread
    private fun registerHandlers() {
        _helper.registerHandlers()
    }

    @AnyThread
    private fun deregisterHandlers() {
        _helper.deregisterHandlers()
    }

    override val isLoaded: Boolean
        @AnyThread get() = _isLoaded.get()

    @AnyThread
    override fun load() {
        Thread.runOnMainThread {
            _logger.debug("$kTag: ${this::load.name}: id = $_adId")
            val showCallback = object : FullScreenContentCallback() {
                override fun onAdShowedFullScreenContent() {
                    Thread.runOnMainThread {
                        _logger.debug("$kTag: ${this::onAdShowedFullScreenContent.name}: id = $_adId")
                    }
                }

                override fun onAdFailedToShowFullScreenContent(error: AdError) {
                    Thread.runOnMainThread {
                        _logger.debug("$kTag: ${this::onAdFailedToShowFullScreenContent.name}: id = $_adId message = ${error.message}")
                        _isLoaded.set(false)
                        _ad = null
                        _bridge.callCpp(_messageHelper.onFailedToShow, ErrorResponse(error.code, error.message).serialize())
                    }
                }

                override fun onAdDismissedFullScreenContent() {
                    Thread.runOnMainThread {
                        _logger.debug("$kTag: ${this::onAdDismissedFullScreenContent.name}: id = $_adId")
                        _isLoaded.set(false)
                        _ad = null
                        _bridge.callCpp(_messageHelper.onClosed)
                    }
                }
            }

            val onAdPaid = object : OnPaidEventListener {
                override fun onPaidEvent(adValue: AdValue) {
                    if (_ad == null) {
                        return;
                    }

                    _onAdPaid(
                        AdPaidResponse(
                            _ad!!.adUnitId,
                            "App Open",
                            adValue.valueMicros,
                            _ad!!.responseInfo.loadedAdapterResponseInfo
                        )
                    );
                }
            }

            val loadCallback = object : AppOpenAdLoadCallback() {
                override fun onAdLoaded(ad: AppOpenAd) {
                    Thread.runOnMainThread {
                        _logger.debug("$kTag: ${this::onAdLoaded.name}: id = $_adId")
                        ad.fullScreenContentCallback = showCallback
                        ad.onPaidEventListener = onAdPaid
                        _isLoaded.set(true)
                        _ad = ad
                        _bridge.callCpp(_messageHelper.onLoaded)
                    }
                }

                override fun onAdFailedToLoad(error: LoadAdError) {
                    Thread.runOnMainThread {
                        _logger.debug("$kTag: ${this::onAdFailedToLoad.name}: id = $_adId message = ${error.message} response = ${error.responseInfo ?: ""}")
                        _bridge.callCpp(_messageHelper.onFailedToLoad, ErrorResponse(error.code, error.message).serialize())
                    }
                }
            }
            val orientation = if (Utils.isLandscape(_application))
                AppOpenAd.APP_OPEN_AD_ORIENTATION_LANDSCAPE else
                AppOpenAd.APP_OPEN_AD_ORIENTATION_PORTRAIT
            AppOpenAd.load(_application, _adId, AdRequest.Builder().build(), orientation, loadCallback)
        }
    }

    @AnyThread
    override fun show() {
        Thread.runOnMainThread {
            _logger.debug("$kTag: ${this::show.name}: id = $_adId")
            val ad = _ad
            if (ad == null) {
                _bridge.callCpp(_messageHelper.onFailedToShow, ErrorResponse(-1, "Null ad").serialize())
                return@runOnMainThread
            }
            val activity = _activity
            if (activity == null) {
                _bridge.callCpp(_messageHelper.onFailedToShow, ErrorResponse(-1, "Null activity").serialize())
                return@runOnMainThread
            }
            ad.show(activity)
        }
    }
}

File: src/ios/ee/ad_mob/AdMobAppOpenAd.swift
//
//  AdMobAppOpenAd.swift
//  ee-x-d542b565
//
//  Created by eps on 12/14/20.
//

import GoogleMobileAds

private let kTag = "\(AdMobAppOpenAd.self)"

internal class AdMobAppOpenAd: NSObject, IFullScreenAd, GADFullScreenContentDelegate {
    private let _bridge: IMessageBridge
    private let _logger: ILogger
    private let _adId: String
    private let _messageHelper: MessageHelper
    private var _helper: FullScreenAdHelper?
    private var _isLoaded = false
    private var _ad: GADAppOpenAd?
    private let _onAdPaid: (AdPaidResponse) -> (Void)

    init(_ bridge: IMessageBridge,
         _ logger: ILogger,
         _ adId: String,
         _ onAdPaid: @escaping (AdPaidResponse) -> (Void)
    ) {
        _bridge = bridge
        _logger = logger
        _adId = adId
        _onAdPaid = onAdPaid
        _messageHelper = MessageHelper("AdMobAppOpenAd", _adId)
        super.init()
        _helper = FullScreenAdHelper(_bridge, self, _messageHelper)
        registerHandlers()
    }

    func destroy() {
        deregisterHandlers()
    }

    func registerHandlers() {
        _helper?.registerHandlers()
    }

    func deregisterHandlers() {
        _helper?.deregisterHandlers()
    }

    var isLoaded: Bool {
        return _isLoaded
    }

    func load() {
        Thread.runOnMainThread {
            self._logger.debug("\(kTag): \(#function): id = \(self._adId)")
            let orientation = Utils.isLandscape()
                ? UIInterfaceOrientation.landscapeLeft
                : UIInterfaceOrientation.portrait
            GADAppOpenAd.load(withAdUnitID: self._adId,
                              request: GADRequest(),
                              orientation: orientation) { ad, error in
                if let error = error {
                    Thread.runOnMainThread {
                        self._logger.debug("\(kTag): \(#function): failed id = \(self._adId) message = \(error.localizedDescription)")
                        self._bridge.callCpp(self._messageHelper.onFailedToLoad, EEJsonUtils.convertDictionary(toString: [
                            "code": (error as NSError).code,
                            "message": error.localizedDescription
                        ]))
                    }
                } else {
                    Thread.runOnMainThread {
                        self._logger.debug("\(kTag): \(#function): succeeded id = \(self._adId)")
                        self._isLoaded = true
                        self._ad = ad
                        self._ad?.fullScreenContentDelegate = self
                        self._ad?.paidEventHandler = {adValue in
                            let responseInfo = ad?.responseInfo
                            let loadedAdNetworkResponseInfo = responseInfo?.loadedAdNetworkResponseInfo
                            let adSourceName = loadedAdNetworkResponseInfo?.adSourceName
                            
                            let adResponse = AdPaidResponse(adUnitId: self._adId, adFormat: "App Open", valueMicros: adValue.value, networkName: adSourceName ?? "")
                            self._onAdPaid(adResponse)
                        }
                        self._bridge.callCpp(self._messageHelper.onLoaded)
                    }
                }
            }
        }
    }

    func show() {
        Thread.runOnMainThread {
            self._logger.debug("\(kTag): \(#function): id = \(self._adId)")
            guard let ad = self._ad else {
                assert(false, "Ad is not initialized")
                self._bridge.callCpp(self._messageHelper.onFailedToShow, EEJsonUtils.convertDictionary(toString: [
                    "code": -1,
                    "message": "Null ad"
                ]))
                return
            }
            guard let rootView = Utils.getCurrentRootViewController() else {
                assert(false, "Current rootView is null")
                self._bridge.callCpp(self._messageHelper.onFailedToShow, EEJsonUtils.convertDictionary(toString: [
                    "code": -1,
                    "message": "Null root view"
                ]))
                return
            }
            ad.present(fromRootViewController: rootView)
        }
    }

    func ad(_ ad: GADFullScreenPresentingAd, didFailToPresentFullScreenContentWithError error: Error) {
        Thread.runOnMainThread {
            self._logger.debug("\(kTag): \(#function): id = \(self._adId) message = \(error.localizedDescription)")
            self._isLoaded = false
            self._ad = nil
            self._bridge.callCpp(self._messageHelper.onFailedToShow, EEJsonUtils.convertDictionary(toString: [
                "code": (error as NSError).code,
                "message": error.localizedDescription
            ]))
        }
    }
    
    func adWillPresentFullScreenContent(_ ad: GADFullScreenPresentingAd) {
        Thread.runOnMainThread {
            self._logger.debug("\(kTag): \(#function): id = \(self._adId)")
        }
    }

    func adDidDismissFullScreenContent(_ ad: GADFullScreenPresentingAd) {
        Thread.runOnMainThread {
            self._logger.debug("\(kTag): \(#function): id = \(self._adId)")
            self._isLoaded = false
            self._ad = nil
            self._bridge.callCpp(self._messageHelper.onClosed)
        }
    }
}
